graphcore_cloud_tools/paperspace_utils/symlink_datasets_and_caches.py

```diff
@@ -401,18 +411,56 @@ def copy_graphcore_s3(args):
     symlink_config = json.loads(json_data)
     datasets = read_gradient_settings(args.gradient_settings_file)
     prepare_cred()
-    _, errors = parallel_download_dataset_from_s3(
-        datasets,
-        symlink_config,
-        max_concurrency=args.max_concurrency,
-        num_concurrent_downloads=args.num_concurrent_downloads,
-        symlink=not args.no_symlink,
-        endpoint_fallback=args.public_endpoint,
-    )
-    if errors:
+
+    max_retries = args.max_retries
+    failed_files = []
+    all_attempts_errors = {}
+
+    for attempt in range(1 + max_retries):  # Including the initial attempt
+        _, errors, failed_files = parallel_download_dataset_from_s3(
+            datasets,
+            symlink_config,
+            max_concurrency=args.max_concurrency,
+            num_concurrent_downloads=args.num_concurrent_downloads,
+            symlink=not args.no_symlink,
+            endpoint_fallback=args.public_endpoint,
+            failed_files=failed_files,
+        )
+
+        if errors:
+            # add and label errors from different attempts
+            if attempt == 0:
+                all_attempts_errors = errors
+            else:
+                key_suffix = f"_retry_{attempt}"
+                for key, value in errors.items():
+                    all_attempts_errors[f"{key}{key_suffix}"] = value
+
+            if failed_files:
+                # retry download failed files
+                if attempt < max_retries:
+                    print(
+                        f"Retry attempt {attempt+1}/{max_retries}: Retrying download for {len(failed_files)} failed files..."
+                    )
+                else:
+                    print(f"All {max_retries} retries exhausted. Failed to download {len(failed_files)} files.")
+            else:
+                # hit a different error, needs special addressing, not necessarily retry download. raise error below.
+                break
+        else:
+            # Successful download. Remove any "failed files" errors from previous unsuccessful attempts.
+            # There may still be other errors like missing datasets that will be raised
+            if any(key.startswith("failed_file_downloads") for key in all_attempts_errors):
```

Comment:
If the name of that key changes this logic will break, is it tested in the test suite?